### DISCUSSION TITLE
Expand file

### DISCUSSION THREAD

**Discussion by 1392-github** [NationalSecurityAgency/ghidra#7162] [Q&A] [answered]
how to expand DLL/SO/EXE file and add address to expanded space?

**Comment by ryanmkurtz** [comment]
You do this with the Memory Map.  There is a `+` button:

![image](https://github.com/user-attachments/assets/cc47528b-38b0-4f1e-98d9-3d4dfad8b8da)

**Reply by 1392-github** [reply]
I tried this already, but when I actually export the dll/so/exe file, that area is not recorded

**Reply by ryanmkurtz** [reply]
![image](https://github.com/user-attachments/assets/a98f8d30-f5df-42eb-8277-87bf349d5cd4)

**Reply by 1392-github** [reply]
Is there any way to do this?

**Reply by ryanmkurtz** [reply]
No, not with Ghidra.

**Reply by 1392-github** [reply]
So why is this feature there?

**Reply by 1392-github** [reply]
So are there other similar methods?

**Reply by ryanmkurtz** [reply]
It wasn't that long ago that we didn't even support exporting the program back out to a file.  When we started storing the original file in the program database, we realized it would be relatively simple to do.  However, if you modify the memory map, you would realistically need to fix up the headers, move other sections around, etc.  It just became a bit out of scope for what Ghidra fundamentally was designed to do.

**Reply by ryanmkurtz** [reply]
> So why is this feature there?

The primary purpose of the feature is get your original program back out of Ghidra.  You can make simple modifications to the binary by editing data and instructions in-place.  I have used this a couple of times to alter the behavior of an existing program.

**Reply by ryanmkurtz** [reply]
> So are there other similar methods?

You'd have to explain what your goals are exactly for me to offer any advice.

**Reply by 1392-github** [reply]
For example, adding new assembly code to a program.

**Reply by ryanmkurtz** [reply, marked as answer]
If you want to do that in Ghidra, you need to clear out some existing file-backed code, and puts yours in its place.  You just have to get creative about where you store it.  Keep in mind, this is not what Ghidra was intended to do, so you'll have to work for it a bit.

**Reply by 1392-github** [reply]
Are there other tools to add new areas? (Excluding paid programs such as IDA)

**Reply by ryanmkurtz** [reply]
Not that i'm aware of, but I am not up to date on what other tools are out there.  We'd have to do the same googling at this point.  Sorry!

**Reply by Wall-AF** [reply]
> > So why is this feature there?
> 
> The primary purpose of the feature is get your original program back out of Ghidra. You can make simple modifications to the binary by editing data and instructions in-place. I have used this a couple of times to alter the behavior of an existing program.

Need more double register support for things like numbers that are larger than 1 register, proper segment:offset recognition, etc.
